pkg/azure/minter.go

```diff
@@ -0,0 +1,240 @@
+	}
+
+	roleAssignmentsClient := authorization.NewRoleAssignmentsClient(subscriptionID)
+	roleAssignmentsClient.Authorizer = rmAuthorizer
+
+	roleDefinitionClient := authorization.NewRoleDefinitionsClient(subscriptionID)
+	roleDefinitionClient.Authorizer = rmAuthorizer
+
+	return &AzureCredentialsMinter{
+		appClient:             addapclient,
+		spClient:              spClient,
+		tenantID:              tenantID,
+		subscriptionID:        subscriptionID,
+		roleAssignmentsClient: roleAssignmentsClient,
+		roleDefinitionClient:  roleDefinitionClient,
+		logger:                logger,
+	}, nil
+}
+
+func (credMinter *AzureCredentialsMinter) CreateOrUpdateAADApplication(ctx context.Context, addAppName string, regenClientSecret bool) (*graphrbac.Application, string, error) {
+	appResp, err := credMinter.appClient.List(ctx, fmt.Sprintf("displayName eq '%v'", addAppName))
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to list AAD applications: %v", err)
+	}
+
+	appItems := appResp.Values()
+	switch len(appItems) {
+	case 0:
+		credMinter.logger.Infof("Creating AAD application %q", addAppName)
+		secret := uuid.NewV4().String()
+		app, err := credMinter.appClient.Create(ctx, graphrbac.ApplicationCreateParameters{
+			DisplayName:             to.StringPtr(addAppName),
+			AvailableToOtherTenants: to.BoolPtr(false),
+			IdentifierUris:          &[]string{"http://localhost/" + uuid.NewV4().String()},
+			PasswordCredentials: &[]graphrbac.PasswordCredential{
+				{
+					Value: &secret,
+					// INFO(jchaloup): Is one year enough?
+					// Should we also prolong the end date or generate new password in case it's outdated?
+					EndDate: &date.Time{Time: time.Now().AddDate(1, 0, 0)},
+				},
+			},
+		})
+		if err != nil {
+			return nil, "", fmt.Errorf("unable to create AAD application: %v", err)
+		}
+		return &app, secret, nil
+	case 1:
+		credMinter.logger.Infof("Found AAD application %q", addAppName)
+		clientSecret := ""
+		if regenClientSecret {
+			secret := uuid.NewV4().String()
+			_, err := credMinter.appClient.UpdatePasswordCredentials(ctx, *appItems[0].ObjectID, graphrbac.PasswordCredentialsUpdateParameters{
+				Value: &[]graphrbac.PasswordCredential{
+					{
+						Value:   &secret,
+						EndDate: &date.Time{Time: time.Now().AddDate(1, 0, 0)},
+					},
+				},
+			})
+			if err != nil {
+				return nil, "", err
+			}
+			clientSecret = secret
+		}
+		return &appItems[0], clientSecret, nil
+	default:
+		return nil, "", fmt.Errorf("found more than 1 AAD application with %q name, will do nothing", addAppName)
+	}
+}
+
+func (credMinter *AzureCredentialsMinter) CreateOrUpdateServicePrincipal(ctx context.Context, appID string) (*graphrbac.ServicePrincipal, error) {
+	spResp, err := credMinter.spClient.List(ctx, fmt.Sprintf("appId eq '%v'", appID))
+	if err != nil {
+		return nil, err
+	}
+
+	spItems := spResp.Values()
+	switch len(spItems) {
+	case 0:
```

Comment:
It looks like both of these functions will handle the app/SP being deleted in Azure outside of the operator, which is great. :+1: